### PR TITLE
C51-368: Calculate Activity Count for Case from Backend

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this PR, the Activity Count Calculation for Case has been moved to backend.
Also performance fixes has been made so that the Manage Cases page loads faster.

## Technical Details
1. **category_count**
Now the count of activities in different categories is completely done from the backend. Previously to count `overdue` activities, All Activities of a case needed to be fetched by frontend. And this resulted in, fetching a large data set which means loading is more slow.
A new function `calculate_activities_for_category` is created in the backend, which helps to calculate the values.

Also the structure of `category_count` has been changed to
```
{ 
  category_count: {
    tasks: { complete: <number>, incomplete: <number>, overdue: <number> },
    communication: { complete: <number>, incomplete: <number>, overdue: <number> },
    <all other categories>,
    none: { complete: <number>, incomplete: <number>, overdue: <number> },
  }
}
```
and on the frontend, all the partials has been updated to use this new structure.

2. `countIncompleteOtherTasks()` function is updated to follow new structure.

3. From `get-case-query-params.factory.js`, removed parameters to fetch all activities.
Also renamed all Api calls to appropriate names
For example: `api.Activity.get.2` has been renamed to `api.Activity.get.recentCommunication`

4. In `get-case-query-params.factory.js`, new API calls has been added to fetch Scheduled activities. And the counts has been moved from `category_count` to `status_count` object.

5. To reduce unwanted network calls while fetching cases, in `case-list-table.directive.js`
    * The watcher has been set to `page.num` instead of `page`. Because changing `page.size` or `page.total` should not fetch new cases.
    * Also the following line has been changed to stop updating ` $scope.page.num` unnecessarily, which in return fires the watcher.
     `$scope.page.num = result[0].page || $scope.page.num;` 

      changed to 

      ```
       if (result[0].page) {
         $scope.page.num = result[0].page;
       }
      ```
6. After making the above fix(point 5), the contact Icons did not show in case cards. Because when Contact Cards were initialising, the Contacts Cache does not have the data about the contacts. To fix this, Contacts Cache now emits an event `civicase::contacts-cache::contacts-added` whenever new contacts are added, and the icons are refetched when this event is fired.

### CODE REVIEW

File: ang/civicase/case/contact-case-tab/directives/contact-case-tab-case-details.directive.html [hunk 1]
```diff
@@ -113,7 +113,7 @@
         <span>
           <span
             class="civicase__contact-card civicase__contact-card--manager"
-            civicase-contact-card contacts="[item.manager]">
+            contact-card contacts="[item.manager]">
```
Comment (reneolivo) [inline ang/civicase/case/contact-case-tab/directives/contact-case-tab-case-details.directive.html]: are we removing the prefix from all directives?

File: api/v3/Case/Getdetails.php [hunk 2]
```diff
@@ -235,6 +240,39 @@ function civicrm_api3_case_getdetails($params) {
...
+  GROUP BY ca.case_id";
+
+  $dao = CRM_Core_DAO::executeQuery($query);
+
+  while ($dao->fetch()) {
+    $categoryName = empty($category) ? 'none' : $category;
+    $statusTypeName = $isOverdue ? "overdue" : $statusType;
+    $result['values'][$dao->case_id]['category_count'][$categoryName][$statusTypeName] = (int) $dao->count;
+  }
+}
```
Comment (reneolivo) [inline api/v3/Case/Getdetails.php]: I'm ok with the solution, but I'm worried that we are fetching all this data even for cases that are not going to display the count (like the ones in the list table). Can you confirm this for me?

File: ang/civicase/case/contact-case-tab/directives/contact-case-tab-case-details.directive.html (hunk 1 above)
Comment (deb1990) [reply ang/civicase/case/contact-case-tab/directives/contact-case-tab-case-details.directive.html]: Sorry this was removed by accident, fixed.

File: api/v3/Case/Getdetails.php (hunk 2 above)
Comment (deb1990) [reply api/v3/Case/Getdetails.php]: As discussed, this is fine.

**Comment by deb1990** [comment]
Also fixed broken layout in Contact Case tab
Before
2019-03-12 at 6 39 PM

After
2019-03-12 at 6 37 PM

**Pull request merged**